File: X_API_NetCore_Master/X.Util.Extend/Core/CacheContext.cs
using System;
using X.Util.Entities;
using X.Util.Entities.Enum;
using X.Util.Entities.Interface;

namespace X.Util.Extend.Core
{
    public class CacheContext<TResult, TChannel> : IContext<TResult, TChannel>
    {
        private readonly IProvider<TChannel> _channel;
        public readonly CacheOptions Options;
        public readonly Func<TResult, bool> CallSuccess;

        public CacheContext(IProvider<TChannel> channel, Func<TResult, bool> callSuccess, CacheOptions options)
        {
            _channel = channel;
            Options = options;
            CallSuccess = callSuccess ?? (t => true);
        }

        public IProvider<TChannel> Channel => _channel;

        private Func<CacheResult<TResult>> Convert(Func<TResult> caller)
        {
            return () =>
            {
                var iresult = caller();
                return new CacheResult<TResult>
                {
                    Succeed = CallSuccess(iresult),
                    Result = iresult
                };
            };
        }

        private static Func<TResult> Convert(Func<CacheResult<TResult>> caller)
        {
            return () =>
            {
                var iresult = caller();
                return iresult != null ? iresult.Result : default;
            };
        }


        public Func<TResult> Calling(ActionContext<TResult> context, Func<TResult> caller)
        {
            switch (Options.CacheExpireType)
            {
                case EnumCacheExpireType.Absolute:
                    return Convert(() => CoreCache.Default.GetAbsoluteCacheData(Convert(caller),
                        context.Request.Method,
                        context.Request.ActionArguments,
                        Options.CacheAppVersion,
                        Options.DebugWithoutCache,
                        Options.AddContext,
                        Options.CacheType,
                        Options.CacheTimeLevel,
                        Options.CacheTimeExpire));
                case EnumCacheExpireType.Sliding:
                    return Convert(() => CoreCache.Default.GetSlidingCacheData(Convert(caller),
                        context.Request.Method,
                        context.Request.ActionArguments,
                        Options.CacheAppVersion,
                        Options.DebugWithoutCache,
                        Options.AddContext,
                        Options.CacheType,
                        Options.CacheTimeLevel,
                        Options.CacheTimeExpire));
            }
            return caller;
        }

        public void Called(ActionContext<TResult> context) { }

        public void OnException(ActionContext<TResult> context) { }

        public int Priority => 0;
    }
}

File: X_Front_Master/x.front.type-components/src/util/extend/cacheContext.js
"use strict";
import { cacheData } from "./cacheData.js";

/**
 * CacheContext
 * @param {any} Provider
 * @param {any} Options:{Type:Object{Key=;Version=;CallSuccess=function{returnType bool;};Date=;}}
 */
function CacheContext(Provider, Options) {
    this.Provider = Provider;
    this.Options = Options;
}

/**
 * Calling
 * @param {any} context:Context.getActionContext
 * @param {any} caller
 */
CacheContext.prototype.Calling = function(context, caller) {
    let options = this.Options;
    let CallSuccess = options.CallSuccess;
    let args = context.ActionArguments;
    let bindThis = context.ActionBindThis;
    let box = function(caller) {
        return () => {
            let iresult = caller.apply(bindThis, args);
            return {
                Succeed: CallSuccess(iresult),
                Result: iresult
            };
        };
    };
    let unbox = function(caller) {
        return () => {
            var iresult = caller();
            return iresult ? iresult.Result : null;
        };
    }
    return unbox(() => {
        return cacheData.getCacheData(options.Key, options.Version, options.Date, box(caller), args, bindThis);
    });
}

/**
 * Called
 * @param {any} context:Context.getActionContext
 */
CacheContext.prototype.Called = function(context) {
    console.debug("debug:CacheContext-Called;ignore", context);
}

/**
 * OnException
 * @param {any} context:Context.getActionContext
 */
CacheContext.prototype.OnException = function(context) {
    console.debug("debug:CacheContext-OnException;ignore", context);
}

export default CacheContext;